File: dotnet/tests/ApplePurchaseReceiptVerifier.Tests/Fixtures.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Security.Cryptography;
using System.Text;
using ApplePurchaseReceiptVerifier.Internal;

namespace ApplePurchaseReceiptVerifier.Tests;

/// <summary>
/// The fixture registry from <c>fixtures/cases.json</c>: ids to logical bytes,
/// each checked against the SHA-256 the registry records for them.
/// </summary>
internal static class Fixtures
{
    /// <summary>The parsed <c>fixtures/cases.json</c> document.</summary>
    internal static readonly OrderedMap Cases = LoadCases();

    /// <summary>The <c>fixtures/</c> directory, found by walking up from the test binary.</summary>
    internal static readonly string Root = FindFixturesDirectory();

    private static readonly Dictionary<string, byte[]> Cache = new(StringComparer.Ordinal);

    /// <summary>Every registered fixture id.</summary>
    internal static IEnumerable<string> Ids
    {
        get
        {
            foreach (KeyValuePair<string, object?> entry in Registry)
            {
                yield return entry.Key;
            }
        }
    }

    private static OrderedMap Registry =>
        Cases["fixtures"] as OrderedMap ?? throw new InvalidOperationException("cases.json has no fixtures map");

    /// <summary>
    /// The decoded logical bytes of a registered fixture, verified against the
    /// digest the registry records.
    /// </summary>
    /// <remarks>
    /// <c>contentSha256</c> is the anti-drift guarantee for the vectors: a
    /// fixture that is regenerated, re-encoded or silently edited changes the
    /// bytes every port verifies, and the expected fields would then be pinned
    /// to something no other port ever saw. Checking it here is what makes the
    /// guarantee load-bearing rather than documentary — and the digest is over
    /// the logical bytes, which are the bytes handed to the library.
    /// </remarks>
    internal static byte[] Bytes(string id)
    {
        lock (Cache)
        {
            if (Cache.TryGetValue(id, out byte[]? cached))
            {
                return cached;
            }
        }

        if (Registry[id] is not OrderedMap entry)
        {
            throw new InvalidOperationException($"harness error: cases.json registers no fixture \"{id}\"");
        }

        string path = Str(entry, "path");
        string codec = Str(entry, "codec");
        string expected = Str(entry, "contentSha256");
        byte[] raw = File.ReadAllBytes(Path.Combine(Root, path.Replace('/', Path.DirectorySeparatorChar)));
        byte[] bytes = codec switch
        {
            "raw" => raw,
            "base64" => Convert.FromBase64String(Strip(Encoding.ASCII.GetString(raw))),
            "utf8" => Encoding.UTF8.GetBytes(Encoding.UTF8.GetString(raw).Trim()),
            _ => throw new InvalidOperationException($"harness error: unknown fixture codec \"{codec}\""),
        };

        string actual = Hex(SHA256.HashData(bytes));
        if (!string.Equals(actual, expected, StringComparison.Ordinal))
        {
            throw new InvalidOperationException(
                $"fixture \"{id}\" ({path}, codec {codec}) has drifted: cases.json records "
                + $"contentSha256 {expected}, the decoded bytes hash to {actual}");
        }

        lock (Cache)
        {
            Cache[id] = bytes;
        }

        return bytes;
    }

    /// <summary>The fixture's bytes as UTF-8 text — the JWS transport form.</summary>
    internal static string Text(string id) => Encoding.UTF8.GetString(Bytes(id));

    private static string Strip(string text)
    {
        StringBuilder builder = new(text.Length);
        foreach (char c in text)
        {
            if (!char.IsWhiteSpace(c))
            {
                builder.Append(c);
            }
        }

        return builder.ToString();
    }

    private static string Str(OrderedMap map, string key) =>
        map[key] as string ?? throw new InvalidOperationException($"harness error: missing \"{key}\"");

    private static string Hex(byte[] value)
    {
        StringBuilder builder = new(value.Length * 2);
        foreach (byte b in value)
        {
            builder.Append(b.ToString("x2", CultureInfo.InvariantCulture));
        }

        return builder.ToString();
    }

    private static OrderedMap LoadCases()
    {
        // Parsed with the reader the library itself ships, so the conformance
        // run is also a few hundred more inputs through it.
        return Json.ParseObject(File.ReadAllText(Path.Combine(FindFixturesDirectory(), "cases.json")));
    }

    private static string FindFixturesDirectory()
    {
        // Walk up rather than hardcode "../../..": the relative depth changes
        // the moment the target-framework count does.
        DirectoryInfo? directory = new(AppContext.BaseDirectory);
        while (directory is not null)
        {
            string candidate = Path.Combine(directory.FullName, "fixtures");
            if (File.Exists(Path.Combine(candidate, "cases.json")))
            {
                return candidate;
            }

            directory = directory.Parent;
        }

        throw new InvalidOperationException("harness error: could not locate fixtures/cases.json");
    }
}

File: php/tests/Support/Fixtures.php
<?php

declare(strict_types=1);

namespace EminDeniz99\ApplePurchaseReceiptVerifier\Tests\Support;

use RuntimeException;

/**
 * Locates the repository's shared `fixtures/` directory and decodes the
 * fixtures `cases.json` registers, checking each one against the SHA-256 the
 * registry records for its DECODED logical bytes.
 *
 * That digest check is not decoration. It is the only mechanical defence
 * against the whole conformance suite going green against fixture bytes that
 * were regenerated, re-encoded or quietly edited: the pinned expectations
 * would then describe bytes no other port ever saw.
 */
final class Fixtures
{
    /** @var array<string, string> */
    private static array $cache = [];

    /** @var array<string, mixed>|null */
    private static ?array $document = null;

    /**
     * The directory is found by walking up from this file, not by a
     * `../../..` literal, so moving the port one level deeper is a rename
     * rather than a silent breakage.
     */
    public static function directory(): string
    {
        $dir = __DIR__;
        for ($i = 0; $i < 12; ++$i) {
            if (is_file($dir . '/fixtures/cases.json')) {
                return $dir . '/fixtures';
            }
            $parent = dirname($dir);
            if ($parent === $dir) {
                break;
            }
            $dir = $parent;
        }

        throw new RuntimeException('harness error: could not locate fixtures/cases.json by walking up from ' . __DIR__);
    }

    /** @return array<string, mixed> */
    public static function cases(): array
    {
        if (self::$document === null) {
            $json = file_get_contents(self::directory() . '/cases.json');
            if ($json === false) {
                throw new RuntimeException('harness error: cases.json is unreadable');
            }
            /** @var array<string, mixed> $parsed */
            $parsed = json_decode($json, true, 64, JSON_THROW_ON_ERROR);
            self::$document = $parsed;
        }

        return self::$document;
    }

    /** @return array<string, array{path: string, role: string, codec: string, contentSha256: string}> */
    public static function registry(): array
    {
        /** @var array<string, array{path: string, role: string, codec: string, contentSha256: string}> */
        return self::cases()['fixtures'];
    }

    /** The decoded logical bytes of a registered fixture, digest-checked. */
    public static function bytes(string $id): string
    {
        if (isset(self::$cache[$id])) {
            return self::$cache[$id];
        }
        $registry = self::registry();
        if (!isset($registry[$id])) {
            throw new RuntimeException("harness error: cases.json registers no fixture \"{$id}\"");
        }
        $entry = $registry[$id];
        $raw = file_get_contents(self::directory() . '/' . $entry['path']);
        if ($raw === false) {
            throw new RuntimeException("harness error: fixture \"{$id}\" ({$entry['path']}) is unreadable");
        }
        $bytes = match ($entry['codec']) {
            'raw' => $raw,
            'base64' => self::strictBase64($id, $raw),
            'utf8' => trim($raw),
            default => throw new RuntimeException(
                "harness error: unknown fixture codec \"{$entry['codec']}\" for \"{$id}\"",
            ),
        };
        $actual = hash('sha256', $bytes);
        if (!hash_equals($entry['contentSha256'], $actual)) {
            throw new RuntimeException(
                "fixture \"{$id}\" ({$entry['path']}, codec {$entry['codec']}) has drifted: "
                . "cases.json records contentSha256 {$entry['contentSha256']}, "
                . "the decoded bytes hash to {$actual}",
            );
        }

        return self::$cache[$id] = $bytes;
    }

    private static function strictBase64(string $id, string $text): string
    {
        $stripped = preg_replace('/\s+/', '', $text) ?? '';
        $decoded = base64_decode($stripped, true);
        if ($decoded === false) {
            throw new RuntimeException("harness error: fixture \"{$id}\" is not valid base64");
        }

        return $decoded;
    }
}
